File: dm labs/S1/DM 3/24.py
import sys

sys.stdin = open('nextperm.in', 'r')
sys.stdout = open('nextperm.out', 'w')


n = int(input())
per = [int(x) for x in input().split()]
perr = per.copy()

boolissimo = [1] * n
minn = per[-1]
maxx = per[-1]
mini = 0
maxi = mini

for i in range(n - 1, -1, -1):
    boolissimo[perr[i] - 1] = 0
    if perr[i] > minn:
        mini = i
        break
    if perr[i] < minn:
        minn = perr[i]

j = perr[mini] - 2
while j > -1 and boolissimo[j] == 1:
    j -= 1
if j == -1:
    print('0 ' * n)
else:
    boolissimo[j] = 1
    perr[mini] = j + 1
    j = mini + 1
    for i in range(n - 1, -1, -1):
        if boolissimo[i] == 0:
            perr[j] = i + 1
            j += 1
    print(*perr)

boolissimo = [1] * n


for i in range(n - 1, -1, -1):
    boolissimo[per[i] - 1] = 0
    if per[i] < maxx:
        maxi = i
        break
    if per[i] > maxx:
        maxx = per[i]
# print(maxi, minn)
j = per[maxi]
while j < n and boolissimo[j] == 1:
    j += 1
if j == n:
    print('0 ' * n)
else:
    boolissimo[j] = 1
    per[maxi] = j + 1
    j = maxi + 1
    for i in range(n):
        if boolissimo[i] == 0:
            per[j] = i + 1
            j += 1
    print(*per)

File: dm labs/S1/DM 3/24.cpp
#include <bits/stdc++.h>

using namespace std;

void num2per(int num, int n, long long *fast_fact) {
    if (num < 0 || num >= fast_fact[n]) {
        while (n--) {
            cout << "0 ";
        }
        cout << '\n';
        return;
    }
    bool boolissimo[n];
    memset(boolissimo, 0, n);
    for (int col = n - 1; col > 0; col--) {
        int t = (num / fast_fact[col]) % (col + 1);
        int tt = 0;
        while (boolissimo[tt]) tt++;
        while (t--)
            while (boolissimo[++tt]);
        boolissimo[tt] = 1;
        cout << tt + 1 << ' ';
    }
    int tt = 0;
    while (boolissimo[tt]) tt++;
    cout << tt + 1 << '\n';
}

int main() {
//    freopen("nextperm.in", "r", stdin);
//    freopen("nextperm.out", "w", stdout);
    int n;
    cin >> n;
    long long fast_fact[n + 1];
    fast_fact[0] = 1;
    for (int i = 1; i <= n; i++)
        fast_fact[i] = i * fast_fact[i - 1];
    int per[n];
    for (int j = 0; j < n; ++j) {
        cin >> per[j];
    }
    bool boolissimo[n];
    memset(boolissimo, 0, n);
    int ans = 0;
    for (int p = 0; p < n; ++p) {
        int s = 0;
        for (int i = 0; i < per[p] - 1; ++i) {
            s += boolissimo[i] ? 0 : 1;
        }
        boolissimo[per[p] - 1] = 1;
        ans += fast_fact[n - p - 1] * s;
    }
    num2per(ans - 1, n, fast_fact);
    num2per(ans + 1, n, fast_fact);
    return 0;
}
